### PR TITLE
Errors/internal

### LINKED ISSUE #537: Create `errors` library so that we can start using it for internal error handling
Refer - Spec section https://hackmd.io/f0WqseIITIqaqRrmYBgy1w#Internal-Node-errors

## Acceptance criteria
- Implement following set of functions as outlined by the spec
```
func New(err error, args ...interface{}) error
func NewListError(olderr, newerr) error
func Len(err) int 
func NewTypedError(type, err error) error
func OfType(type, err error) bool
```
- Unit tests to make sure that each functions works

### PR DESCRIPTION
**Pull request by vedhavyas**
Fixes #537 

Note: This only adds internal errors. We need to add client and peerError later

### CODE REVIEW

File: utils/events.go [hunk 1]
```diff
@@ -1,9 +1,11 @@
 package utils
 
-import "fmt"
+import (
+	"github.com/centrifuge/go-centrifuge/errors"
+)
 
 // ErrEventNotFound when event is not found and need to retry
-var ErrEventNotFound = fmt.Errorf("event not found")
+var ErrEventNotFound = errors.New("event not found")
```
Comment (vimukthi-git) [inline utils/events.go]: `const` please

File: errors/errors.go [hunk 2]
```diff
@@ -0,0 +1,139 @@
+package errors
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+)
+
+// ErrUnknown is an unknown error type
+var ErrUnknown = New("unknown error")
```
Comment (vimukthi-git) [inline errors/errors.go]: `const` please

File: errors/errors_test.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
+// +build unit
+
+package errors
```
Comment (vimukthi-git) [inline errors/errors_test.go]: 👍

File: errors/errors.go [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+	return &typeError{terr: terr, ctxErr: err}
+}
+
+// TypeError can be implemented by any type error
+type TypeError interface {
+	IsOfType(terr error) bool
+}
+
+// isOfType returns if the err t is of type terr
+func (t *typeError) IsOfType(terr error) bool {
```
Comment (vimukthi-git) [inline errors/errors.go]: isOfType => IsOfType

File: errors/errors.go (hunk 2 above)
Comment (vedhavyas) [reply errors/errors.go]: `const` wont work for error types :)

Comment (vedhavyas) [reply errors/errors.go]: It should be a type not an interface for constants. We can work this around but i'm thinking a better solution

Comment (vedhavyas) [reply errors/errors.go]: Pushed `const` changes

Comment (vimukthi-git) [reply errors/errors.go]: i had the solution to this in the spec, as you may see

File: errors/errors.go [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+func (e Error) Error() string {
+	return string(e)
+}
+
+// New returns a new error with message passed.
+// if args are passed, we will format the message with args
+// Example:
+// New("some error") returns error with exact string passed
+// New("some error: %v", "some context") returns error with message "some error: some context"
+func New(msg string, args ...interface{}) error {
```
Comment (mikiquantum) [inline errors/errors.go]: Can we enforce here the format? Meaning:
```
err := New("some error", "some context")
```
And it is the `New` function that formats it:
```
...
if len(args) > 0 {
		return Error(fmt.Sprintf("%s: %s", msg, args...))
}
```
That way we have the standard built in.

File: errors/errors.go [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+}
+
+// Error returns the error in string
+func (t *typeError) Error() string {
+	return fmt.Sprintf("%v: %v", t.terr, t.ctxErr)
+}
+
+// NewTypeError returns a new error of type typeError
+func NewTypeError(terr, err error) error {
+	if terr == nil {
```
Comment (mikiquantum) [inline errors/errors.go]: `terror` lol

File: errors/errors.go (hunk 2 above)
Comment (vedhavyas) [reply errors/errors.go]: Just noticed! awesome

File: errors/errors.go (hunk 5 above)
Comment (vedhavyas) [reply errors/errors.go]: We can enforce it but leaves with multiple functions like std library. Thinking having a single function can be easier.

Comment (mikiquantum) [reply errors/errors.go]: But we are trying to standardize our usage of errors right? 
In addition, looking at the `msg string` argument, doesn't tell me if it is just a string or a format string, so either we make it clear in the arg name and doc or we enforce it, IMHO.

Comment (vedhavyas) [reply errors/errors.go]: Doc says it already. Not sure what could be the name of arg can be. Do you have a suggestion for arg name?

Comment (vedhavyas) [reply errors/errors.go]: Will update the arg to format as per the std lib

**Pull request merged**